File: nodejs/Todo.js
const fs = require('fs')

class Todo {
  constructor () {
    this.dbFilePath = './../db.json'
    this.dbFileContent = JSON.parse(this.readFile())
  }

  writeFile (fileContent) {
    fileContent = fileContent || JSON.stringify(this.dbFileContent)
    fs.writeFileSync(this.dbFilePath, fileContent, (err) => {
      if (err) throw err
    })
    return true
  }

  readFile () {
    if (fs.existsSync(this.dbFilePath)) {
      return fs.readFileSync(this.dbFilePath, 'utf8')
    } else {
      this.writeFile('{}')
      return '{}'
    }
  }

  addTask (taskTitle) {
    const lastKey = parseInt(Object.keys(this.dbFileContent).pop())
    const newTaskId = lastKey ? (lastKey + 1) : 1
    this.dbFileContent[newTaskId] = { title: taskTitle, done: 0 }
    return this.writeFile() ? `added: ${taskTitle}` : 'error'
  }

  editTask (taskId, taskTitle) {
    if (this.dbFileContent[taskId]) {
      this.dbFileContent[taskId].title = taskTitle
      return this.writeFile() ? `edited: ${taskId}` : 'error'
    }
  }

  markAsDone (taskId) {
    if (this.dbFileContent[taskId]) {
      this.dbFileContent[taskId].done = 1
      return this.writeFile() ? `marked as done: ${taskId}` : 'error'
    }
  }

  markAsUndone (taskId) {
    if (this.dbFileContent[taskId]) {
      this.dbFileContent[taskId].done = 0
      return this.writeFile() ? `marked as undone: ${taskId}` : 'error'
    }
  }

  removeTask (taskId) {
    if (this.dbFileContent[taskId]) {
      delete this.dbFileContent[taskId]
      return this.writeFile() ? `removed: ${taskId}` : 'error'
    }
  }

  listTasks () {
    let tasks = 'ID\tDone\tTask\n===================='
    Object.keys(this.dbFileContent).map((id) => {
      const title = this.dbFileContent[id].title
      const done = this.dbFileContent[id].done ? '[x]' : '[ ]'
      tasks += `\n#${id}\t${done}\t${title}`
    })
    return tasks
  }
}

module.exports = new Todo()

File: python/Todo.py
import os
import json

class Todo:
  def __init__ (self):
    self.dbFilePath = './../db.json'
    self.dbFileContent = json.loads(self.readFile())

  def writeFile (self, fileContent = None):
    fileContent = fileContent if fileContent else json.dumps(self.dbFileContent)
    file = open(self.dbFilePath, 'w')
    file.write(fileContent)
    file.close()
    return True

  def readFile (self):
    if (os.path.exists(self.dbFilePath)):
      file = open(self.dbFilePath, 'r')
      return file.read()
    else:
      self.writeFile('{}')
      return '{}'

  def addTask (self, taskTitle):
    lastKey = int(sorted(self.dbFileContent.keys())[-1]) if self.dbFileContent else 0
    newTaskId = (lastKey + 1) if lastKey else 1
    self.dbFileContent[newTaskId] = { 'title': taskTitle, 'done': 0 }
    return 'added: {0}'.format(taskTitle) if self.writeFile() else 'error'

  def editTask (self, taskId, taskTitle):
    if self.dbFileContent[taskId]:
      self.dbFileContent[taskId]['title'] = taskTitle
      return 'edited: {0}'.format(taskTitle) if self.writeFile() else 'error'

  def markAsDone (self, taskId):
    if self.dbFileContent[taskId]:
      self.dbFileContent[taskId]['done'] = 1
      return 'marked as done: {0}'.format(taskId) if self.writeFile() else 'error'

  def markAsUndone (self, taskId):
    if self.dbFileContent[taskId]:
      self.dbFileContent[taskId]['done'] = 0
      return 'marked as undone: {0}'.format(taskId) if self.writeFile() else 'error'

  def removeTask (self, taskId):
    if self.dbFileContent[taskId]:
      del self.dbFileContent[taskId]
      return 'removed: {0}'.format(taskId) if self.writeFile() else 'error'

  def listTasks (self):
    tasks = 'ID\tDone\tTask\n===================='
    for id in self.dbFileContent:
      title = self.dbFileContent[id]['title']
      done = '[x]' if self.dbFileContent[id]['done'] else '[ ]'
      tasks += '\n#{0}\t{1}\t{2}'.format(id, done, title)
    return tasks

todo = Todo()
